ikarus/io/vtkwriter.hh

```diff
@@ -0,0 +1,249 @@
+  }
+
+  template <typename Basis>
+  constexpr auto dimBasis = []() {
+    // Case 1 PowerBasis
+    if constexpr (requires { Basis::PreBasis::children; })
+      return Basis::PreBasis::children;
+    // Case 2 SubSpaceBasis or Scalar Basis
+    else
+      return 1;
+  }();
+
+  template <typename Container>
+  constexpr auto sizeOfContainer = []() {
+    if constexpr (requires { Container::dimension; })
+      return Container::dimension;
+    else if constexpr (requires { std::tuple_size<Container>::value; })
+      return std::tuple_size<Container>::value;
+    else
+      return 1ul;
+  }();
+
+} // namespace Impl
+
+/**
+ * \struct Writer
+ * \brief Manages writing results using VTK, based on assembler and data collector.
+ *
+ * \tparam AS Type of the assembler.
+ * \tparam DC Type of the data collector.
+ * \tparam Base Base class for VTK writer.
+ */
+template <typename AS, typename DC, typename Base>
+requires(Concepts::FlatAssembler<AS> && Concepts::DataCollector<DC>)
+struct Writer : public Base
+{
+public:
+  using Assembler     = AS;
+  using GridView      = typename Assembler::GridView;
+  using FERequirement = typename Assembler::FERequirement;
+  using FEContainer   = typename Assembler::FEContainer;
+  using FEType        = typename std::remove_cvref_t<FEContainer>::value_type;
+
+  using DataCollector = DC;
+  using VTKWriter     = Base;
+
+  /**
+   * \brief Constructor with assembler and additional arguments.
+   *
+   * \param assembler Shared pointer to assembler.
+   * \param args Additional arguments.
+   */
+  template <class... Args>
+  Writer(std::shared_ptr<AS> assembler, Args... args)
+      : Base(assembler->gridView(), std::forward<Args>(args)...),
+        assembler_(assembler) {}
+
+  /**
+   * \brief Constructor with assembler, data collector reference, and additional arguments.
+   *
+   * \param assembler Shared pointer to assembler.
+   * \param dc Reference to data collector
+   * \param args Additional arguments.
+   */
+  template <class... Args>
+  Writer(std::shared_ptr<AS> assembler, DC& dc, Args... args)
+      : Base(dc, std::forward<Args>(args)...),
+        assembler_(assembler) {}
+
+  /**
+   * \brief Constructor with assembler, data collector, and additional arguments.
+   *
+   * \param assembler Shared pointer to assembler.
+   * \param dc Data collector as a rvalue reference
+   * \param args Additional arguments.
+   */
+  template <class... Args>
+  Writer(std::shared_ptr<AS> assembler, DC&& dc, Args... args)
+      : Base(std::move(dc), std::forward<Args>(args)...),
+        assembler_(assembler) {}
+
+  /**
+   * \brief Adds a result function for the given data tag.
+   *
+   * \tparam RF Type of the result function.
+   * \param resultFunction The Ikarus::ResultFunction.
+   * \param dataTag The data tag.
+   */
+  template <typename RF, Impl::Concepts::DataType DT>
+  void addResultFunction(RF&& resultFunction, DT dataTag) {
```

Comment:
This is not really needed to be a templated. I suggest just create an enum with the two types and then decide at runtime if `addCellData` or `addPointData` being so template heavy is not much of a benefit here.